File: Android/handy/src/main/java/de/paulherter/swiftly/tv/TvErweitert.kt
package de.paulherter.swiftly.tv

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.runtime.snapshots.SnapshotStateList
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.unit.dp
import androidx.compose.material3.Text
import de.paulherter.swiftly.Kopfzeile
import de.paulherter.swiftly.SwiftlyAnwendung
import de.paulherter.swiftly.alsJson
import de.paulherter.swiftly.bereinigt
import de.paulherter.swiftly.gemeinsam.Stil
import de.paulherter.swiftly.gemeinsam.Zeichen
import de.paulherter.swiftly.gemeinsam.uebersetzt
import de.paulherter.swiftly.kern.Kern
import de.paulherter.swiftly.kopfzeilenAus

/**
 * **„Erweitert" auf dem Fernseher** — eigene Header fuer einen Dienst vor dem Server (Issue #4).
 * Vorlage: `Sources/tvOS/TVErweitert.swift`; dieselben Zeilen (`Kopfzeile`) wie auf dem Telefon,
 * mit den Feldern und Knoepfen des Fernsehers: Name und Wert nebeneinander, dahinter das Minus.
 * Zugeklappt und leer, damit sich fuer niemanden sonst etwas aendert. Der Wert ist verdeckt.
 */
@Composable
fun TvErweitert(zeilen: SnapshotStateList<Kopfzeile>, aufgeklappt: Boolean = false, modifier: Modifier = Modifier) {
    var offen by rememberSaveable { mutableStateOf(false) }
    LaunchedEffect(Unit) { if (zeilen.isNotEmpty()) offen = true }
    Column(modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(20.dp)) {
        if (!aufgeklappt) {
            TvKnopf(uebersetzt("Erweitert"), if (offen) Zeichen.WinkelRunter else Zeichen.WinkelRechts,
                    symbolNachText = true) { offen = !offen }
        }
        if (offen || aufgeklappt) {
            Text(uebersetzt("Für einen Dienst vor deinem Server, etwa Cloudflare Access, Authelia oder Pangolin. Swiftly schickt die Header nur an diese Adresse."),
                 style = TvStil.klein, color = Stil.schriftSehrLeise)
            zeilen.forEach { zeile ->
                androidx.compose.runtime.key(zeile.id) {
                    Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
                        Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(20.dp)) {
                            TvFeld(zeile.name, { zeile.name = it }, uebersetzt("Header-Name"), Modifier.weight(1f))
                            TvFeld(zeile.wert, { zeile.wert = it }, uebersetzt("Wert"), Modifier.weight(1f), geheim = true)
                            TvKnopf(null, Zeichen.Minus) { zeilen.removeAll { it.id == zeile.id } }
                        }
                        if (zeile.gesperrt) Text(uebersetzt("Den setzt Swiftly selbst."), style = TvStil.klein, color = Stil.warnung)
                    }
                }
            }
            TvKnopf(uebersetzt("Header hinzufügen"), Zeichen.Plus) { zeilen.add(Kopfzeile()) }
        }
    }
}

/** Die eigenen Header des aktiven Servers bearbeiten — aus Profil → Server. Vorlage `TVEigeneKoepfeSeite`. */
@Composable
fun TvEigeneKoepfeSeite(app: SwiftlyAnwendung, schliessen: () -> Unit) {
    val server = app.aktiverServer()
    val zeilen = remember { mutableStateListOf<Kopfzeile>().apply { addAll(kopfzeilenAus(app.eigeneKoepfe(server))) } }
    var gesichert by remember { mutableStateOf(Kern.eigenkoepfeBereinigt(app.eigeneKoepfe(server))) }
    val fokus = ersterFokus()
    val geaendert = zeilen.bereinigt() != gesichert
    Box(Modifier.fillMaxSize().background(Stil.grund).verticalScroll(rememberScrollState()), contentAlignment = Alignment.TopCenter) {
        Column(Modifier.width(900.dp).padding(vertical = 80.dp)) {
            Text(uebersetzt("Eigene Header"), style = TvStil.titelGross, color = Stil.schrift)
            TvErweitert(zeilen, aufgeklappt = true, modifier = Modifier.padding(top = 44.dp))
            Row(Modifier.padding(top = 44.dp), horizontalArrangement = Arrangement.spacedBy(20.dp)) {
                if (geaendert && server != null) {
                    TvKnopf(uebersetzt("Sichern")) {
                        app.eigeneKoepfeSichern(zeilen.alsJson(), server)
                        gesichert = Kern.eigenkoepfeBereinigt(app.eigeneKoepfe(server))
                        schliessen()
                    }
                }
                TvKnopf(uebersetzt(if (geaendert) "Abbrechen" else "Fertig"), modifier = Modifier.focusRequester(fokus)) { schliessen() }
            }
        }
    }
}

File: Sources/tvOS/TVErweitert.swift
import JellyfinKit
import SwiftUI

/// **„Erweitert" auf dem Fernseher** — eigene Header für einen Dienst vor dem
/// Server (Issue #4). Die Vorlage ist `Shared/Erweitertbereich.swift`; hier
/// dieselben Zeilen (``Kopfzeile``) mit den Feldern und Knöpfen des
/// Fernsehers. Zugeklappt und leer, damit sich für niemanden sonst etwas
/// ändert. Der Wert ist verdeckt wie ein Passwort.
struct TVErweitert: View {
    @Binding var zeilen: [Kopfzeile]
    var aufgeklappt = false

    @State private var offen = false

    var body: some View {
        VStack(alignment: .leading, spacing: 20) {
            if !aufgeklappt {
                Button {
                    withAnimation(Stil.einblenden) { offen.toggle() }
                } label: {
                    HStack(spacing: 12) {
                        Text("Erweitert")
                        Image(systemName: offen ? "chevron.down" : "chevron.right")
                    }
                }
                .buttonStyle(KnopfStil())
            }
            if offen || aufgeklappt { inhalt }
        }
        .onAppear { if !zeilen.isEmpty { offen = true } }
    }

    private var inhalt: some View {
        VStack(alignment: .leading, spacing: 20) {
            Text("Für einen Dienst vor deinem Server, etwa Cloudflare Access, Authelia oder Pangolin. Swiftly schickt die Header nur an diese Adresse.")
                .font(Stil.klein)
                .foregroundStyle(Stil.schriftSehrLeise)
                .fixedSize(horizontal: false, vertical: true)

            ForEach($zeilen) { $zeile in
                VStack(alignment: .leading, spacing: 12) {
                    HStack(spacing: 20) {
                        Eingabefeld(platzhalter: "Header-Name", text: $zeile.name)
                        Eingabefeld(platzhalter: "Wert", text: $zeile.wert, sicher: true)
                        Button {
                            zeilen.removeAll { $0.id == zeile.id }
                        } label: {
                            Image(systemName: "minus")
                        }
                        .buttonStyle(KnopfStil(nurSymbol: true))
                        .accessibilityLabel(Text("Entfernen"))
                    }
                    if zeile.gesperrt {
                        Text("Den setzt Swiftly selbst.")
                            .font(Stil.klein)
                            .foregroundStyle(Stil.warnung)
                    }
                }
            }

            Button {
                zeilen.append(Kopfzeile())
            } label: {
                HStack(spacing: 12) {
                    Image(systemName: "plus")
                    Text("Header hinzufügen")
                }
            }
            .buttonStyle(KnopfStil())
        }
    }
}

/// Die eigenen Header des aktiven Servers bearbeiten — aus Profil → Server.
struct TVEigeneKoepfeSeite: View {
    let model: AppModel
    let schliessen: () -> Void

    @State private var zeilen: [Kopfzeile] = []
    @State private var gesichert: [Eigenkopf] = []

    private var server: URL? { model.session?.serverURL }

    var body: some View {
        ZStack(alignment: .topLeading) {
            Stil.grund.ignoresSafeArea()
            ScrollView {
                VStack(alignment: .leading, spacing: 0) {
                    Text("Eigene Header")
                        .font(Stil.titelGross)
                        .foregroundStyle(Stil.schrift)
                    TVErweitert(zeilen: $zeilen, aufgeklappt: true)
                        .padding(.top, 44)
                    HStack(spacing: 20) {
                        if zeilen.koepfe != gesichert {
                            Button("Sichern") {
                                guard let server else { return }
                                model.eigeneKoepfeSichern(zeilen.koepfe, fuer: server)
                                gesichert = model.eigeneKoepfe(fuer: server)
                                schliessen()
                            }
                            .buttonStyle(KnopfStil())
                        }
                        Button(zeilen.koepfe != gesichert ? "Abbrechen" : "Fertig", action: schliessen)
                            .buttonStyle(KnopfStil())
                    }
                    .padding(.top, 44)
                }
                .frame(width: 900, alignment: .leading)
                .frame(maxWidth: .infinity)
                .padding(.vertical, 80)
            }
        }
        .onAppear {
            gesichert = model.eigeneKoepfe(fuer: server)
            zeilen = Kopfzeile.aus(gesichert)
        }
    }
}
